### PR TITLE
Cache root component metadata creation

### PR DESCRIPTION
**Pull request by melix**
This pull request introduces caching of root component metadata in configurations. There are several benefits:

- root component metadata creation can be very expensive. The more configurations there are, the slower it is, so caching allows us to reduce the overhead
- the creation is now externalized, meaning there's nothing specific to a `Configuration`, allowing us to reuse the component in different configurations of the same project
- if a configuration is resolved several times, the same root component metadata is reused

Caching makes sure that we don't cache if any of the configurations have been mutated in between. This PR, however, does **not** introduce laziness in the root component metadata, so we still eagerly build all configuration metadata for the whole component.

See gradle/performance#532

### CODE REVIEW

**Comment by melix** [comment]
Reading the report [here](https://builds.gradle.org/repository/download/Gradle_Branches_Performance_PerformanceTestCoordinatorLinux/2800735:id/results/performance/build/performance-tests/scenario-report.html) it seems to have a measureable impact on all "Android" test cases. For the other tests, it's hard to tell. I guess the difference comes from the fact Android projects have significantly more configurations than usual projects.

**Comment by oehme** [comment]
It's because Android 2.3 resolves all Configurations in all Projects. It's a particular stress test for this metadata creation.

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/moduleconverter/DefaultRootComponentMetadataBuilderTest.groovy [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        builder.validator.validateMutation(mutationType)
+        def otherRoot = builder.toRootComponentMetaData()
+
+        then:
+        otherRoot.is(root) == cached
+
+        where:
+        mutationType                                | cached
+        MutationValidator.MutationType.DEPENDENCIES | false
+        MutationValidator.MutationType.ARTIFACTS    | true
```
Comment (oehme) [inline subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/moduleconverter/DefaultRootComponentMetadataBuilderTest.groovy]: ❌ Why can we cache in this case?

Comment (melix) [inline subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/moduleconverter/DefaultRootComponentMetadataBuilderTest.groovy]: We should not, you're right.

Comment (melix) [inline subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/moduleconverter/DefaultRootComponentMetadataBuilderTest.groovy]: fixed

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/moduleconverter/DefaultRootComponentMetadataBuilderTest.groovy [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        when:
+        builder.validator.validateMutation(mutationType)
+        def otherRoot = builder.toRootComponentMetaData()
+
+        then:
+        otherRoot.is(root) == cached
+
+        where:
+        mutationType                                | cached
+        MutationValidator.MutationType.DEPENDENCIES | false
```
Comment (oehme) [inline subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/moduleconverter/DefaultRootComponentMetadataBuilderTest.groovy]: ❌ What about attributes and the like?

Comment (melix) [inline subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/moduleconverter/DefaultRootComponentMetadataBuilderTest.groovy]: Attributes are locked during the root component metadata creation. It's only unfortunate that we allow mutation of configurations _after_ root component metadata has been created. It should be an error, but for backwards compatibility, we cannot enforce it like we did for attributes.

Comment (oehme) [inline subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/moduleconverter/DefaultRootComponentMetadataBuilderTest.groovy]: Okay, thanks for the explanation.

**Pull request merged**